lib/cyrusdb_twom.c

```diff
@@ -0,0 +1,2596 @@
+    }
+
+    header->flags
+        = ntohl(*((uint32_t *)(base + OFFSET_FLAGS)));
+
+    // XXX - check flags for other comparison engines?
+    file->compar = bsearch_ncompare_raw;
+
+    header->generation
+        = ntohll(*((uint64_t *)(base + OFFSET_GENERATION)));
+
+    header->num_records
+        = ntohll(*((uint64_t *)(base + OFFSET_NUM_RECORDS)));
+
+    header->num_commits
+        = ntohll(*((uint64_t *)(base + OFFSET_NUM_COMMITS)));
+
+    header->dirty_size
+        = ntohll(*((uint64_t *)(base + OFFSET_DIRTY_SIZE)));
+
+    header->repack_size
+        = ntohll(*((uint64_t *)(base + OFFSET_REPACK_SIZE)));
+
+    header->current_size
+        = ntohll(*((uint64_t *)(base + OFFSET_CURRENT_SIZE)));
+
+    header->maxlevel
+        = ntohl(*((uint32_t *)(base + OFFSET_MAXLEVEL)));
+
+    if (db->nocsum) return 0;
+
+    uint32_t csum = ntohl(*((uint32_t *)(base + OFFSET_CSUM)));
+    if (file->csum(base, OFFSET_CSUM) != csum) {
+        xsyslog(LOG_ERR, "DBERROR: twom header checksum failure",
+                         "filename=<%s>",
+                         db->fname);
+        return CYRUSDB_IOERROR;
+    }
+
+    return 0;
+}
+
+static size_t tm_roundup(size_t offset)
+{
+    size_t page_size = 1<<14; // 16k
+    return ((offset + offset / 4) + page_size - 1) & ~(page_size - 1);
+}
+
+/* we keep open ONLY the current file 
+ * and any others which still have a reference to them */
+
+static void _remove_txn(struct txn **ptr)
+{
+    struct txn *this = *ptr;
+    struct txn *next = this->next;
+    this->file->refcount--;
+    free(this);
+    *ptr = next;
+}
+
+static void _remove_file(struct tm_file **ptr)
+{
+    struct tm_file *this = *ptr;
+    struct tm_file *next = (*ptr)->next;
+    assert(!this->refcount);
+    assert(!this->has_lock);
+    if (this->base) munmap(this->base, this->size);
+    if (this->fd != -1) close(this->fd);
+    free(this);
+    *ptr = next;
+}
+
+static void empty_db(struct dbengine *db)
+{
+    if (db->loc.file) {
+        db->loc.file->refcount--;
+        db->loc.file = NULL;
+    }
+    while (db->write_txn)
+        _remove_txn(&db->write_txn);
+    while (db->read_txn)
+        _remove_txn(&db->read_txn);
+    while (db->openfile)
+        _remove_file(&db->openfile);
+}
+
+static void tm_cleanup(struct dbengine *db)
+{
+    // close any open files with a zero refcount (no loc or txn
+    // still points there) except the first one
+    if (!db->openfile) return;
+    struct tm_file **ptr = &db->openfile->next;
+    while (*ptr) {
+        if (!(*ptr)->refcount) {
+            _remove_file(ptr);
+        }
+        else {
+            ptr = &((*ptr)->next);
+        }
+    }
+}
+
+/* end cleanup */
+
+static inline int tm_commit(struct dbengine *db, size_t len)
+{
+    assert(db->openfile);
+    if (!db->openfile->dirty) return 0;
+    assert(db->openfile->has_lock == 2);
+    db->openfile->dirty = 0;
+    return msync(db->openfile->base, len, MS_SYNC|MS_INVALIDATE);
```

Comment:
What is tm_commit's purpose? My understanding of msync is that it exists to make sure data is flushed to disk (or at least... to the filesystem layer). I think the only think it protects us from is power loss / server crash and reboot. Is that what we're aiming for here?

What is the MS_INVALIDATE for? With MAP_SHARED, all processes using this mmap'd region should always only see the updated data, I think MS_INVALIDATE only affects things reading the file from the OS or not using MAP_SHARED?

Reply:
Any time we msync with MS_SYNC we risk blocking on file IO I think. Is that okay?

Reply:
Yes, that's the point of this - tm_commit's purpose is "make sure this stuff is on disk".  It's part of the 3 commit safety:
1) mark header dirty; then sync
2) make changes which are all revertable by recovery1 in case of an abort or crash; then sync
3) mark header clean; then sync

So the atomic bit is just (3), and it's only a write of 96 bytes at the top of the file, so should always be within a page.


I should read more about MS_INVALIDATE and see if it's necessary.

Reply:
https://stackoverflow.com/questions/60547532/whats-the-exact-meaning-of-the-flag-ms-invalidate-in-msync - says it's basically a NOOP (as is MS_ASYNC), but making things posixly correct seems sensible.